snowplow-common-internal/kinesis/src/main/scala/com/snowplowanalytics/snowplow/sources/kinesis/KinesisSource.scala

```diff
@@ -0,0 +1,237 @@
+      if (x.sequenceNumber > y.sequenceNumber) x else y
+  }
+
+  private def kinesisCheckpointer[F[_]: Parallel: Sync]: KinesisCheckpointer[F] = new KinesisCheckpointer[F] {
+    def combine(x: Map[String, KinesisMetadata[F]], y: Map[String, KinesisMetadata[F]]): Map[String, KinesisMetadata[F]] =
+      x |+| y
+
+    val empty: Map[String, KinesisMetadata[F]] = Map.empty
+    def ack(c: Map[String, KinesisMetadata[F]]): F[Unit] =
+      c.values
+        .toList
+        .parTraverse_ { metadata =>
+          metadata.ack
+            .recoverWith {
+              case _: ShutdownException =>
+                // The ShardRecordProcessor instance has been shutdown. This just means another KCL
+                // worker has stolen our lease. It is expected during autoscaling of instances, and is
+                // safe to ignore.
+                Logger[F].warn(s"Skipping checkpointing of shard ${metadata.shardId} because this worker no longer owns the lease")
+
+              case _: IllegalArgumentException if metadata.isLastInShard =>
+                // See https://github.com/snowplow/enrich/issues/657
+                // This can happen at the shard end when KCL no longer allows checkpointing of the last record in the shard.
+                // We need to release the semaphore, so that fs2-aws handles checkpointing the end of the shard.
+                Logger[F].warn(
+                  s"Checkpointing failed on last record in shard. Ignoring error and instead try checkpointing of the shard end"
+                ) *>
+                  Sync[F].delay(metadata.lastRecordSemaphore.release())
+
+              case _: IllegalArgumentException if metadata.lastRecordSemaphore.availablePermits === 0 =>
+                // See https://github.com/snowplow/enrich/issues/657 and https://github.com/snowplow/enrich/pull/658
+                // This can happen near the shard end, e.g. the penultimate batch in the shard, when KCL has already enqueued the final record in the shard to the fs2 queue.
+                // We must not release the semaphore yet, because we are not ready for fs2-aws to checkpoint the end of the shard.
+                // We can safely ignore the exception and move on.
+                Logger[F].warn(
+                  s"Checkpointing failed on a record which was not the last in the shard. Meanwhile, KCL has already enqueued the final record in the shard to the fs2 queue. Ignoring error and instead continue processing towards the shard end"
+                )
+            }
+        }
+    def nack(c: Map[String, KinesisMetadata[F]]): F[Unit] = Applicative[F].unit
+  }
+
+  private def kinesisStream[F[_]: Async](config: KinesisSourceConfig): Stream[F, LowLevelEvents[Map[String, KinesisMetadata[F]]]] = {
+    val resources =
+      for {
+        region <- config.region.fold(KinesisSourceConfig.getRuntimeRegion.fold(Resource.raiseError[F, Region, Throwable], Resource.pure[F, Region]))(Resource.pure)
```

Comment:
This is.... ok.

It will work, and that's what I care about most.  So I don't want to complain too loudly.

But I will give some opinions:

First, I find it very hard to read anything that uses `option.fold(default)(f)` if `default` and `f` are themselves very long to read.  My eyes just can't scan it easily.  And code legibility is important, because your primary audience is your colleagues (the compiler is only the secondary audience).

IMO if something is hard to read, then it's worth a refactor out of kindness to your reviewer 😉

Next.....

This will cause an exception if the user forgot to configure `region` in their config file.  But that's just a config mistake, so it should be caught at config-parsing time.  So the `ConfigParser` module can give a kind message to the user explaining what went wrong and where.

I think that's a nicer experience for the user instead of getting an exception when the app tries to connect to the stream.

In other words... I would make `region` non-optional in the `KinesisSourceConfig` class.  And `getRuntimeRegion` can be part of the decoder.

Reply:
appreciate this feedback, thanks